Nightly Cuda_Pthread and Pthread build failures

It looks like following merge of #3861 some builds with Pthreads enabled began failing, apparently when gcc is host compiler (intel builds with Pthreads passed):

cuda/9.2 with Cuda_Pthreads enabled has error messages of type:

```
06:46:16 cd /home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/TestAll_2021-04-09_06.02.03/cuda/9.2/Cuda_Pthread-release/core/unit_test && /home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/bin/nvcc_wrapper  -DGTEST_HAS_PTHREAD=0 -DGTEST_HAS_TR1_TUPLE=0 -DKOKKOS_DEPENDENCE -I/home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/tpls/gtest -I/home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/TestAll_2021-04-09_06.02.03/cuda/9.2/Cuda_Pthread-release/core/unit_test -I/home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/core/unit_test -I/home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/core/unit_test/category_files -I/home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/TestAll_2021-04-09_06.02.03/cuda/9.2/Cuda_Pthread-release -I/home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/TestAll_2021-04-09_06.02.03/cuda/9.2/Cuda_Pthread-release/core/src -I/home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/core/src -I/home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/TestAll_2021-04-09_06.02.03/cuda/9.2/Cuda_Pthread-release/containers/src -I/home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/containers/src -I/home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/TestAll_2021-04-09_06.02.03/cuda/9.2/Cuda_Pthread-release/algorithms/src -I/home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/algorithms/src  -O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -O3 -DNDEBUG   -arch=sm_35 -Xcudafe --diag_suppress=esa_on_defaulted_function_ignored -pthread -std=c++14 -o CMakeFiles/KokkosCore_UnitTest_Threads.dir/threads/TestThreads_TeamBasic.cpp.o -c /home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/TestAll_2021-04-09_06.02.03/cuda/9.2/Cuda_Pthread-release/core/unit_test/threads/TestThreads_TeamBasic.cpp
06:46:16 nvcc_wrapper - *warning* you have set multiple optimization flags (-O*), only the last is used because nvcc can only accept a single optimization setting.
06:46:16 /home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/core/src/Threads/Kokkos_ThreadsTeam.hpp(116): warning: calling a __host__ function("Kokkos::Impl::spinwait_while_equal<int> ") from a __host__ __device__ function("Kokkos::Impl::ThreadsExecTeamMember::team_fan_in const") is not allowed
06:46:16 
06:46:16 /home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/core/src/Threads/Kokkos_ThreadsTeam.hpp(123): warning: calling a __host__ function("Kokkos::Impl::spinwait_while_equal<int> ") from a __host__ __device__ function("Kokkos::Impl::ThreadsExecTeamMember::team_fan_in const") is not allowed
06:46:16 
06:46:16 /home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/core/src/Threads/Kokkos_ThreadsTeam.hpp(933): error: identifier "Test::VectorScanReducer::checkScan< ::Kokkos::Threads, ( ::Test::VectorScanReducer::ScanType)1,  ::Kokkos::Prod<double,  ::Kokkos::Threads> > ::n_vector_range" is undefined in device code
06:46:16 
06:46:16 /home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/core/src/Threads/Kokkos_ThreadsTeam.hpp(933): error: identifier "Test::VectorScanReducer::checkScan< ::Kokkos::Threads, ( ::Test::VectorScanReducer::ScanType)0,  ::Kokkos::Prod<double,  ::Kokkos::Threads> > ::n_vector_range" is undefined in device code
06:46:16 
06:46:16 /home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/core/src/Threads/Kokkos_ThreadsTeam.hpp(933): error: identifier "Test::VectorScanReducer::checkScan< ::Kokkos::Threads, ( ::Test::VectorScanReducer::ScanType)1,  ::Kokkos::Max<double,  ::Kokkos::Threads> > ::n_vector_range" is undefined in device code
06:46:16 
06:46:16 /home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/core/src/Threads/Kokkos_ThreadsTeam.hpp(933): error: identifier "Test::VectorScanReducer::checkScan< ::Kokkos::Threads, ( ::Test::VectorScanReducer::ScanType)0,  ::Kokkos::Max<double,  ::Kokkos::Threads> > ::n_vector_range" is undefined in device code
06:46:16 
06:46:16 /home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/core/src/Threads/Kokkos_ThreadsTeam.hpp(933): error: identifier "Test::VectorScanReducer::checkScan< ::Kokkos::Threads, ( ::Test::VectorScanReducer::ScanType)1,  ::Kokkos::Min<double,  ::Kokkos::Threads> > ::n_vector_range" is undefined in device code
06:46:16 
06:46:16 /home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_cuda_92/kokkos/core/src/Threads/Kokkos_ThreadsTeam.hpp(933): error: identifier "Test::VectorScanReducer::checkScan< ::Kokkos::Threads, ( ::Test::VectorScanReducer::ScanType)0,  ::Kokkos::Min<double,  ::Kokkos::Threads> > ::n_vector_range" is undefined in device code
```

Reproducer instructions - cuda/9.2 with Cuda_Pthread
```
         module load sems-env sems-cmake/3.17.1 sems-cuda/9.2 sems-gcc/6.1.0

         $KOKKOS_PATH/generate_makefile.bash --with-devices=Cuda,Pthread --arch=Kepler35 --compiler=$KOKKOS_PATH/bin/nvcc_wrapper --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="14" --with-cuda --kokkos-path=$KOKKOS_PATH --no-examples --with-options=enable_large_mem_tests
```

gcc/7.3.0 Pthreads build (linker issues - undefined references):
```
05:50:05 CMakeFiles/KokkosCore_UnitTest_Threads.dir/threads/TestThreads_TeamBasic.cpp.o: In function `Test::VectorScanReducer::checkScan<Kokkos::Threads, (Test::VectorScanReducer::ScanType)1, Kokkos::Prod<double, Kokkos::Threads> >::TeamThreadRangeFunctor::operator()(unsigned long) const':
05:50:05 /home/jenkins/slave/workspace/Kokkos_develop_kokkosdev2_spotcheck-dbg/kokkos/core/unit_test/TestTeamVector.hpp:916: undefined reference to `Test::VectorScanReducer::checkScan<Kokkos::Threads, (Test::VectorScanReducer::ScanType)1, Kokkos::Prod<double, Kokkos::Threads> >::n_vector_range'
05:50:05 CMakeFiles/KokkosCore_UnitTest_Threads.dir/threads/TestThreads_TeamBasic.cpp.o: In function `Test::VectorScanReducer::checkScan<Kokkos::Threads, (Test::VectorScanReducer::ScanType)0, Kokkos::Prod<double, Kokkos::Threads> >::TeamThreadRangeFunctor::operator()(unsigned long) const':
05:50:05 /home/jenkins/slave/workspace/Kokkos_develop_kokkosdev2_spotcheck-dbg/kokkos/core/unit_test/TestTeamVector.hpp:916: undefined reference to `Test::VectorScanReducer::checkScan<Kokkos::Threads, (Test::VectorScanReducer::ScanType)0, Kokkos::Prod<double, Kokkos::Threads> >::n_vector_range'
05:50:05 CMakeFiles/KokkosCore_UnitTest_Threads.dir/threads/TestThreads_TeamBasic.cpp.o: In function `Test::VectorScanReducer::checkScan<Kokkos::Threads, (Test::VectorScanReducer::ScanType)1, Kokkos::Max<double, Kokkos::Threads> >::TeamThreadRangeFunctor::operator()(unsigned long) const':
05:50:05 /home/jenkins/slave/workspace/Kokkos_develop_kokkosdev2_spotcheck-dbg/kokkos/core/unit_test/TestTeamVector.hpp:916: undefined reference to `Test::VectorScanReducer::checkScan<Kokkos::Threads, (Test::VectorScanReducer::ScanType)1, Kokkos::Max<double, Kokkos::Threads> >::n_vector_range'
05:50:05 CMakeFiles/KokkosCore_UnitTest_Threads.dir/threads/TestThreads_TeamBasic.cpp.o: In function `Test::VectorScanReducer::checkScan<Kokkos::Threads, (Test::VectorScanReducer::ScanType)0, Kokkos::Max<double, Kokkos::Threads> >::TeamThreadRangeFunctor::operator()(unsigned long) const':
05:50:05 /home/jenkins/slave/workspace/Kokkos_develop_kokkosdev2_spotcheck-dbg/kokkos/core/unit_test/TestTeamVector.hpp:916: undefined reference to `Test::VectorScanReducer::checkScan<Kokkos::Threads, (Test::VectorScanReducer::ScanType)0, Kokkos::Max<double, Kokkos::Threads> >::n_vector_range'
05:50:05 CMakeFiles/KokkosCore_UnitTest_Threads.dir/threads/TestThreads_TeamBasic.cpp.o: In function `Test::VectorScanReducer::checkScan<Kokkos::Threads, (Test::VectorScanReducer::ScanType)1, Kokkos::Min<double, Kokkos::Threads> >::TeamThreadRangeFunctor::operator()(unsigned long) const':
05:50:05 /home/jenkins/slave/workspace/Kokkos_develop_kokkosdev2_spotcheck-dbg/kokkos/core/unit_test/TestTeamVector.hpp:916: undefined reference to `Test::VectorScanReducer::checkScan<Kokkos::Threads, (Test::VectorScanReducer::ScanType)1, Kokkos::Min<double, Kokkos::Threads> >::n_vector_range'
05:50:05 CMakeFiles/KokkosCore_UnitTest_Threads.dir/threads/TestThreads_TeamBasic.cpp.o: In function `Test::VectorScanReducer::checkScan<Kokkos::Threads, (Test::VectorScanReducer::ScanType)0, Kokkos::Min<double, Kokkos::Threads> >::TeamThreadRangeFunctor::operator()(unsigned long) const':
05:50:05 /home/jenkins/slave/workspace/Kokkos_develop_kokkosdev2_spotcheck-dbg/kokkos/core/unit_test/TestTeamVector.hpp:916: undefined reference to `Test::VectorScanReducer::checkScan<Kokkos::Threads, (Test::VectorScanReducer::ScanType)0, Kokkos::Min<double, Kokkos::Threads> >::n_vector_range'
05:50:05 collect2: error: ld returned 1 exit status
```

Reproducer instructions
```
06:40:28         module load sems-env sems-cmake/3.17.1 sems-gcc/7.3.0
06:40:28 
06:40:28         $KOKKOS_PATH/generate_makefile.bash --with-devices=Pthread --arch=SNB --compiler=g++ --cxxflags="-g -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized " --cxxstandard="14" --kokkos-path=$KOKKOS_PATH --with-options=disable_deprecated_code --no-examples  --debug
```

--- a/core/unit_test/TestTeamVector.hpp
+++ b/core/unit_test/TestTeamVector.hpp
@@ -867,12 +867,11 @@ class TestTripleNestedReduce {
 namespace VectorScanReducer {
 enum class ScanType : bool { Inclusive, Exclusive };
 
-template <typename ExecutionSpace, ScanType scan_type, class Reducer>
+template <typename ExecutionSpace, ScanType scan_type, int n,
+          int n_vector_range, class Reducer>
 struct checkScan {
-  static constexpr int n              = 1000000;
-  static constexpr int n_vector_range = 100;
-  const int n_team_thread_range       = 1000;
-  const int n_per_team                = n_team_thread_range * n_vector_range;
+  const int n_team_thread_range = 1000;
+  const int n_per_team          = n_team_thread_range * n_vector_range;
 
   using size_type  = typename ExecutionSpace::size_type;
   using value_type = typename Reducer::value_type;
@@ -977,10 +976,6 @@ struct checkScan {
       ASSERT_EQ(host_outputs(i), expected(i));
   }
 };
-template <typename ExecutionSpace, ScanType scan_type, class Reducer>
-constexpr int checkScan<ExecutionSpace, scan_type, Reducer>::n;
-template <typename ExecutionSpace, ScanType scan_type, class Reducer>
-constexpr int checkScan<ExecutionSpace, scan_type, Reducer>::n_vector_range;
 }  // namespace VectorScanReducer
 
 #if !(defined(KOKKOS_IMPL_CUDA_CLANG_WORKAROUND) || defined(KOKKOS_ENABLE_HIP))
@@ -1024,24 +1019,27 @@ TEST(TEST_CATEGORY, parallel_scan_with_reducers) {
   using T = double;
   using namespace VectorScanReducer;
 
-  checkScan<TEST_EXECSPACE, ScanType::Exclusive,
+  static constexpr int n              = 1000000;
+  static constexpr int n_vector_range = 100;
+
+  checkScan<TEST_EXECSPACE, ScanType::Exclusive, n, n_vector_range,
             Kokkos::Prod<T, TEST_EXECSPACE>>()
       .run();
-  checkScan<TEST_EXECSPACE, ScanType::Inclusive,
+  checkScan<TEST_EXECSPACE, ScanType::Inclusive, n, n_vector_range,
             Kokkos::Prod<T, TEST_EXECSPACE>>()
       .run();
 
-  checkScan<TEST_EXECSPACE, ScanType::Exclusive,
+  checkScan<TEST_EXECSPACE, ScanType::Exclusive, n, n_vector_range,
             Kokkos::Max<T, TEST_EXECSPACE>>()
       .run();
-  checkScan<TEST_EXECSPACE, ScanType::Inclusive,
+  checkScan<TEST_EXECSPACE, ScanType::Inclusive, n, n_vector_range,
             Kokkos::Max<T, TEST_EXECSPACE>>()
       .run();
 
-  checkScan<TEST_EXECSPACE, ScanType::Exclusive,
+  checkScan<TEST_EXECSPACE, ScanType::Exclusive, n, n_vector_range,
             Kokkos::Min<T, TEST_EXECSPACE>>()
       .run();
-  checkScan<TEST_EXECSPACE, ScanType::Inclusive,
+  checkScan<TEST_EXECSPACE, ScanType::Inclusive, n, n_vector_range,
             Kokkos::Min<T, TEST_EXECSPACE>>()
       .run();
 }

--- a/core/unit_test/TestTeamVector.hpp
+++ b/core/unit_test/TestTeamVector.hpp
@@ -977,6 +977,10 @@ struct checkScan {
       ASSERT_EQ(host_outputs(i), expected(i));
   }
 };
+template <typename ExecutionSpace, ScanType scan_type, class Reducer>
+constexpr int checkScan<ExecutionSpace, scan_type, Reducer>::n;
+template <typename ExecutionSpace, ScanType scan_type, class Reducer>
+constexpr int checkScan<ExecutionSpace, scan_type, Reducer>::n_vector_range;
 }  // namespace VectorScanReducer
 
 #if !(defined(KOKKOS_IMPL_CUDA_CLANG_WORKAROUND) || defined(KOKKOS_ENABLE_HIP))

